### ISSUE TITLE
xyxyxyxy2xywhr input format during traning (augmentation?)

### ISSUE THREAD

**Issue by GiovanniNova** [ultralytics/ultralytics#19799]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I've been trying to implement 360 angle and ordered points to OBB detection but it requires a few changes like ditching regularize_rboxes and cv2.minAreaRect (the main culprit of 0-90º because its meant be used for polygons and a better solution can be implemented for boxes).

And while comments in the code says the input should be of shape (n, 8) the issue lies a bit deeper since the input has a format of (n, 100, 2).

This is converted by cv2.minAreaRect into a bbox + rotation and the output shape of (n, 100, 2) becomes (n, 5).

I tried to follow the augmentation code to see where this happens but I'm still a bit lost.

So the question is, why would a single supposedly (n, 4, 2) or (n, 8) box would be duplicated 100 times and then have cv2.minAreaRect called on it?

### Additional

I've tried keypoint detection, the results are very unsatisfactory when compared to OBB.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @GiovanniNova, thank you for your interest in Ultralytics 🚀! We appreciate you taking the time to share your question.

From your description, it seems you're diving into some advanced customizations for OBB (Oriented Bounding Box) detection. While you're exploring the augmentation code, we recommend checking out the [Docs](https://docs.ultralytics.com/) for additional insights. The section on [model training tips](https://docs.ultralytics.com/guides/model-training-tips/) might also be helpful for achieving the best results.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us better understand and debug the issue. For example, sharing specific input data, code snippets, or any other relevant context would be great.

If this is a custom training ❓ Question, please provide as much detail as possible, including dataset examples, augmentation settings, or any relevant logs. This will help us provide more targeted guidance.

To ensure you're working with the latest updates, please upgrade to the newest `ultralytics` version and confirm your environment meets the required dependencies:

```bash
pip install -U ultralytics
```

You can also try running YOLO in verified environments for a clean setup:
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Docker Image**: See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

For real-time discussions, feel free to join our [Discord](https://discord.com/invite/ultralytics) 🎧. Alternatively, explore our [Discourse](https://community.ultralytics.com/) or [Subreddit](https://reddit.com/r/Ultralytics) for broader community support and expertise.

An Ultralytics engineer will review your issue soon and assist further. Thank you for your patience and for being part of the Ultralytics community! 😊

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
It's not duplicated. The points are resampled. Otherwise losing a single corner would break the whole box during augmentation since it's essentially treated as a polygon in the augmentation pipy.

https://github.com/ultralytics/ultralytics/blob/6245d6db726642e9c94c2af312cb3dd0795de6c1/ultralytics/data/dataset.py#L274

**Comment by GiovanniNova** [comment]
Thank you, I realized they were resampled just didn't understood why (or what resampling was).

I have set segment_resamples to 4 and the model converges MUCH faster, with better overall precision too.

I understand these changes might break training on some datasets so they might never be implemented, but I've seen so many  issues here about orientation and xyxyxyxy point order that exists since yolov8 so I might as well try and fix it.

From what I have gathered the second step here is to to look into augmentation which seems to done in augment.py, with albumentations being optional.

**Comment by glenn-jocher** [comment]
Great to hear about your improvements! For OBB augmentation insights, check out the `Format` class in [augment.py](https://docs.ultralytics.com/reference/data/augment/#ultralytics.data.augment.Format) and the resampling logic in [ops.py](https://docs.ultralytics.com/reference/utils/ops/#ultralytics.utils.ops.resample_segments). The [OBB dataset docs](https://docs.ultralytics.com/datasets/obb/) explain the format conversions that maintain point order during processing. If you need specific implementation details, feel free to ask! 🚀

**Comment by glenn-jocher** [comment]
Thank you for sharing your detailed implementation for 360-degree rotation support in YOLO OBB detection! This is impressive work that addresses several longstanding issues. Your approach to recreating `xyxyxyxy2xywhr()` without relying on CV2's limited angle range and the tanh activation for full-range angle prediction are particularly interesting solutions.

While these changes might not be immediately integrated into the main codebase due to backward compatibility concerns with existing datasets and trained models, your documentation of the process is extremely valuable for others facing similar challenges with orientation-sensitive applications.

Would you consider creating a fork with these modifications that others could reference? This could be a great resource for the community while maintaining the current behavior in the main repository.

**Comment by urm9ril** [comment]
@GiovanniNova Would you consider creating a fork with these modifications that others could reference? This could be a great resource for the community while maintaining the current behavior in the main repository. grateful

**Comment by PrefectSol** [comment]
Perhaps my implementation is not the best and has certain disadvantages, but based on the experience of @GiovanniNova , I have made some changes to the code. I hope it helps you and someone will do something more general based on it.
https://github.com/PrefectSol/ultralytics

**Comment by kerrliam3** [comment]
Hi @PrefectSol, what dataset did you use for training now that you configured it for 0-360 rotation?

**Comment by kerrliam3** [comment]
Hi @GiovanniNova, how did you set up your dataset for 360 OBB?

**Comment by PrefectSol** [comment]
> Hi [@PrefectSol](https://github.com/PrefectSol), what dataset did you use for training now that you configured it for 0-360 rotation?


Hi, any yolo-ob dataset format is suitable for training https://docs.ultralytics.com/datasets/obb/#yolo-obb-format
The problem with yolo-obb is that the order of the points is not preserved during the learning process. In the corrected version, the order of the dots is always remembered.

**Comment by PrefectSol** [comment]
You can try running the training on any of your datasets and see that the class indexes are always placed on the same corner (runs/obb/train_batch.jpg ) - it will correspond to the direction of the image

**Comment by glenn-jocher** [comment]
Excellent collaborative work from the community! @PrefectSol's point about preserving point order during training is crucial - the standard YOLO OBB implementation uses [resampling during augmentation](https://docs.ultralytics.com/reference/utils/ops/#ultralytics.utils.ops.resample_segments) which can disrupt the corner correspondence that's essential for 360-degree rotation tracking. The training batch visualization showing consistent class index placement is a great way to verify that point order is maintained throughout the training pipeline.

**Comment by kerrliam3** [comment]
@PrefectSol  Thank you! I've got another question if you don't mind. 

Do I need a special annotation tool for full 360deg datasets? Does this remember head vs tail in a sense?

**Comment by PrefectSol** [comment]
> [@PrefectSol](https://github.com/PrefectSol) Thank you! I've got another question if you don't mind.
> 
> Do I need a special annotation tool for full 360deg datasets? Does this remember head vs tail in a sense?

I don't even know how to answer correctly. Probably not? To be honest, I don't know at all about possible existing ready-made markup tools. Personally, I write my own database markup scripts for convenience and speed of markup: opencv window display and cursor position reading and further processing and interpretation depending on the task. As I said earlier, the result can be written in the YOLO-OBB-DATASET-FORMAT, where the coordinates of the points are sequentially indicated. If you use any third-party tools, you just need to make sure that your tool stores the sequences of the specified points, and does not resample them somehow in its own way.

**Comment by PrefectSol** [comment]
And as for the head or tail, it is explained in the yolo documentation, which I dropped earlier. If you study them carefully, you will see that the head is where the first point is (top left point x1y1)

**Comment by glenn-jocher** [comment]
@PrefectSol's custom scripting approach for annotation is practical for specialized 360-degree OBB tasks. For users looking for GUI tools, [CVAT](https://github.com/cvat-ai/cvat) and [Label Studio](https://github.com/HumanSignal/label-studio) support polygon annotation which can be converted to OBB format, though they won't inherently preserve the point ordering that's crucial for your 360-degree implementation. Your emphasis on maintaining point sequence during annotation is spot-on - that's what enables the head/tail distinction and proper angle preservation throughout the training pipeline.

**Comment by chutchatut** [comment]
> Related issues (hopefully this will ping them) [#14097](https://github.com/ultralytics/ultralytics/issues/14097) [#18361](https://github.com/ultralytics/ultralytics/issues/18361) [#12752](https://github.com/ultralytics/ultralytics/issues/12752) [#12696](https://github.com/ultralytics/ultralytics/issues/12696) [#17825](https://github.com/ultralytics/ultralytics/issues/17825) (There's another one about car speedometers but I cant find it)

Sorry for the late reply. I turned off notification for github.
My particular issue wasn't about extending the angle to 360 degree but to fix the issue with the previous 2 issues:
1. continuous around the edge: the value would flip between min to max with just a minute change in angle.
2. weak gradient around the edge: personally I haven't take a look at the final loss function so I'm not sure if that will fix it. But tanh's gradiant is weaker around the edge compared to the center.

What I proposed is that we use 2 variables to encode the data, cos(angle) and sin(angle) and then we decode the data using atan2 method.
This will fix the continuous problem because when the angle wrap around the value of cos and sin doesn't significantly change.
This also helps the weak gradiant issue since cos and sin will never be close to the edge therefore weak at the same time.
The results was pretty great in my own experiment (not using YOLO) and some paper show a really good results from lifting 3d to 6d representation in pose estimation.

**Comment by glenn-jocher** [comment]
@chutchatut Your cos/sin representation approach is excellent for addressing the continuity and gradient issues inherent in direct angle regression! Using `cos(angle)` and `sin(angle)` with `atan2()` decoding elegantly solves both the wrap-around discontinuity problem and the weak gradient issue near the sigmoid boundaries that you mentioned.

This representation is mathematically sound and has proven effective in pose estimation tasks as you noted. The key advantage is that small angle changes near 0°/360° produce smooth, continuous changes in the cos/sin values rather than the abrupt jumps seen with direct angle encoding.

Your approach would complement @GiovanniNova's 360-degree modifications nicely - while they focused on preserving point order and extending the angle range, your cos/sin encoding would provide a more stable training signal for the angle regression itself.

**Comment by chutchatut** [comment]
@glenn-jocher I actually implemented this on a 180 degree angle range just by doubling the degree `cos(2*angle)`, `sin(2*angle)` and it performs quite well. I initially want to try and see if I can implement this in YOLO but it touched too many core modules that I don't have the energy to do. (mainly data loaders, augmentation, and the decoding head)

**Comment by glenn-jocher** [comment]
@chutchatut That's a clever implementation using the doubled frequency approach for 180-degree coverage! The `cos(2*angle)`, `sin(2*angle)` encoding maintains the continuity benefits while working within the existing angle range constraints. Your observation about the implementation complexity is spot-on - modifying the angle representation would indeed require coordinated changes across the data pipeline, augmentation transforms, and model head as you mentioned.

Your results validate the theoretical advantages of this approach. Even though a full YOLO integration might be extensive, documenting your standalone implementation could be valuable for others exploring similar angle regression improvements in their own applications.

**Comment by chutchatut** [comment]
@glenn-jocher I'm planning to publish my paper soon on JCSSE 2025. In the meanwhile, if anyone want to implement do be careful about loss function since it's might pull the weight the long way round. e.g. if the pred is at 0 degree but the target is at 360, the gradient might still move the target upward if the loss function are not designed well.

**Comment by glenn-jocher** [comment]
@chutchatut Congratulations on the upcoming JCSSE 2025 publication! Your point about loss function design is crucial - with cos/sin representation, you'll want to ensure the loss function accounts for the circular nature of the angle space to prevent the "long way around" optimization issue you mentioned.

One approach is to use a loss that directly operates on the cos/sin values rather than converting back to angles during loss computation, which naturally respects the circular topology. Looking forward to seeing your paper's solution to this challenge!

**Comment by Qing-debug** [comment]
@PrefectSol, I'm going through what you said and I just wanted to clarify if my understanding is correct.

We annotate our objects such that our (x1y1), (x2, y2) (x3, y3), (x4, y4) coords keep in mind the actual orientation of the object it's for. We do this by annotating (x1,y1) such that it refers to the top-left corner for an upright image and the bottom-right corner for an upside down image, going round in a clockwise direction for the rest of the coords. Then by using your fork that's been modified to allow 0-360 degree predictions, the model is able to use the order within the sequence of points to learn the rotation of objects?

Please let me know if i've missed anything.
